subprojects/functional/src/main/java/org/gradle/internal/collect/AtomicHashSet.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.collect;
+
+import io.usethesource.capsule.Set;
+
+import java.util.Iterator;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class AtomicHashSet<T> implements Iterable<T> {
+
+    private AtomicReference<Set.Immutable<T>> set = new AtomicReference<>(Set.Immutable.of());
+
+    public boolean add(T value) {
+        boolean[] added = new boolean[1];
```

Comment:
updateAndGet returns the new value. Do you need a temp boolean array then?
Would you rather save the original value by calling set.get()?

Reply:
It would be great to get rid of the heap allocation here but my understanding is that another thread could still win and modify the collection after the original value has been saved before `updateAndGet`.

Reply:
Right you are

Reply:
So an alternative could be like `set.getAndUpdate(set -> set.__insert(value)).contains(value)` (`getAndUpdate` returns the value `before the update`, so if `value` was not there, then `__insert` did insert the value), however, calling `contains` for each `add` is likely worse than allocating `boolean[1]`.